### PR TITLE
Add eBPF connection tracking, without dependencies on kernel headers

### LINKED ISSUE #1168: walking /proc to obtain connections is very expensive
@tomwilkie suggests we should replace the existing logic with something eBPF based like https://github.com/iovisor/bcc/blob/master/tools/tcpaccept.py and https://github.com/iovisor/bcc/blob/master/tools/tcpconnect.py

### LINKED ISSUE #1260: Short-lived connections not tracked for containers in shared networking namespaces
We currently can only track short-lived connections (using conntrack) for containers which have a unique IP.

This is because  the `/proc/` walker is not fast enough to track short-lived connections, for which we use conntrack. However, conntrack doesn't provide PIDs (just IPs and ports), which is not good enough to narrow down individual processes. Instead, we obtain the IPs from conntrack and match them against the networking interfaces of the containers, which must have a unique IP to be properly attributed. 

In other words, when there's a short-lived connection from/to an IP shared by multiple containers (which happens when they share the networking namespace) we don't know the specific container the connection is from/to. Instead of showing edges to/from all those containers (which would be inaccurate), we decide not to show any edges at all.

Although this is by design, I am tagging it as a bug because that's how an end-user would see it (the design explanation is too low-level).

A path forward would be to use ebpf to track connections, as proposed in #1168 .

### LINKED ISSUE #1961: Remove runtime kernel header dependency from ebpf tracker
In production, a runtime dependency on kernel headers are not acceptable. We should remove it.

Provided a small footprint, one option is to get ebpf bytecode for all minor kernel versions (or even more smartly, for kernel versions API changes).

### LINKED ISSUE #1962: Port ebpf tracker to Go
We don't want to depend on bcc and Python to run the ebpf connection tracker.

We will need a wrapper library on the low level ebpf system calls and probably replicate some of the bcc features in Go (maybe even create a bcc Go backend)

### PR DESCRIPTION
**Pull request by iaguis**
Based on work from Lorenzo, updated by Iago, Alban, and Alessandro

This PR adds connection tracking using eBPF. This feature is not enabled by default.
For now, you can enable it by launching scope with the following command:
```
sudo ./scope launch --probe.ebpf.connections=true
```
Scope Probe also falls back on the old /proc parsing if eBPF is not working
(e.g. too old kernel, or pre-compiled eBPF program missing for the current kernel).
This patch allows scope to get notified of every connection event, without relying
on the parsing of /proc/$pid/net/tcp{,6} and /proc/$pid/fd/*, and therefore
improve performance.

We vendor https://github.com/kinvolk/gobpf-elf-loader in Scope to load and
receive information from the ebpf programs. The Scope build fetches the pre-compiled
ebpf programs from https://hub.docker.com/r/kinvolk/tcptracer-bpf/
(see https://github.com/kinvolk/tcptracer-bpf)

At the moment, the eBPF programs are pre-compiled for:
- Arch 4.8.11-1
- fedora-24
- debian-testing
- coreos

The ebpf programs use kprobes on the following kernel functions:
- tcp_v4_connect
- tcp_v6_connect
- inet_csk_accept
- tcp_close

It generates "connect", "accept" and "close" events containing the
connection tuple but also pid and netns.
Note: the IPv6 events are not plugged in Scope.

probe/endpoint/ebpf.go maintains the list of connections. Similarly to
conntrack, it also keeps the dead connections for one iteration in order
to report short-lived connections.

The code for parsing /proc/$pid/net/tcp{,6} and /proc/$pid/fd/* is still
there and still used at start-up because eBPF only brings us the events
and not the initial state. However, the /proc parsing for the initial
state is now done in foreground instead of background, via newForegroundReader().

NAT resolution on connections from eBPF works in the same way as it did on connections
from /proc: by using conntrack. One of the two conntrack instances was removed since
eBPF detects short-lived connections.

The Scope Docker image size comparison:
- weaveworks/scope in current master:  22 MB (compressed),  68 MB
  (uncompressed)
- weaveworks/scope with this patchset: 24 MB (compressed), 70 MB
  (uncompressed)

Fixes #1168 (walking /proc to obtain connections is very expensive)
Fixes #1260 (Short-lived connections not tracked for containers in
shared networking namespaces)
Fixes #1962 (Port ebpf tracker to Go)
Fixes #1961 (Remove runtime kernel header dependency from ebpf tracker)

Supersedes #2024

### CODE REVIEW

**Comment by rade** [comment]
> Scope fetches the pre-compiled ebpf program from https://hub.docker.com/r/kinvolk/tcptracer-bpf/
(see https://github.com/kinvolk/tcptracer-bpf)

Perhaps include all the binaries in the scope container image? Yes, it adds bloat, but dynamically downloading binaries (on every launch?) is gross.

**Comment by alban** [comment]
> Perhaps include all the binaries in the scope container image? Yes, it adds bloat, but dynamically downloading binaries (on every launch?) is gross.

@rade It is already in the Scope Docker image. It is done at build-time (implemented in Scope's Makefile).

At the moment, the size of the compiled ebpf programs is very small, I think <50KB uncompressed.
https://hub.docker.com/r/kinvolk/tcptracer-bpf/tags/

**Comment by rade** [comment]
Ah, so when you wrote "Scope fetches" you didn't actually mean "Scope" but "The scope build". Also you wrote "Scope fetches the pre-compiled ebpf *program*" (note the singular). Taking into account your comments, I believe that sentence should have read "The scope build fetches the pre-compiled ebpf programs." Correct?

**Comment by alban** [comment]
@rade correct. Sorry for the impreceise wording. I or @iaguis will edit the PR text.

**Comment by alban** [comment]
I closed the two previous iterations of this work, in favor of this PR. For the record, this was:
- iteration 1: #1967 (ebpf using bcc/python)
- iteration 2: #2024 (ebpf using bcc/gobpf)

**Comment by schu** [comment]
Two graphs comparing the performance between master and this branch (by running our test dialer, #2082):

With up to 4K connections (40 dialer container with 100 connections each):

scope-probe-time-4k-conn

With a pool of 100 containers, each holding a random number of connections between 1 and 20:

scope-probe-time-100-containers-with-up-to-20-conn

File: Makefile [hunk 1]
```diff
@@ -22,15 +22,28 @@ CODECGEN_TARGETS=report/report.codecgen.go render/detailed/detailed.codecgen.go
...
+GO_BUILD_FLAGS+=-pkgdir ~
+else
+GO=env $(GO_ENV) go
+endif
+
+NO_CROSS_COMP=unset GOOS GOARCH
+GO_HOST=$(NO_CROSS_COMP); env $(GO_ENV) go
+WITH_GO_HOST_ENV=$(NO_CROSS_COMP); $(GO_ENV)
 IMAGE_TAG=$(shell ./tools/image-tag)
+EBPF_IMAGE=kinvolk/tcptracer-bpf:semaphore-master-475be63
```
Comment (schu) [inline Makefile]: Maybe change this to pull by digest as labels are not unique, e.g.

```
EBPF_IMAGE=kinvolk/tcptracer-bpf@sha256:e1ec43eaddff2712909fc61e869b088da73d283359eb470f12d01e3d074549c0`
```

**Comment by iaguis** [comment]
Superseded by https://github.com/weaveworks/scope/pull/2135

**Pull request closed without merging**